lib/SIL/Utils/PrunedLiveness.cpp

```diff
@@ -309,40 +227,161 @@ void PrunedLivenessBoundary::visitInsertionPoints(
+    switch (use->getOperandOwnership()) {
+    default:
+      updateForUse(use->getUser(), use->isLifetimeEnding());
+      break;
+    case OperandOwnership::NonUse:
+      break;
+    case OperandOwnership::Borrow:
+      updateForBorrowingOperand(use);
+      break;
+    }
   }
-  llvm_unreachable("No user in LiveWithin block");
 }
 
-void PrunedLivenessBoundary::compute(const PrunedLiveness &liveness) {
-  for (SILBasicBlock *bb : liveness.getDiscoveredBlocks()) {
+template <typename LivenessWithDefs>
+bool PrunedLiveRange<LivenessWithDefs>::isWithinBoundary(
+    SILInstruction *inst) const {
+  assert(asImpl().isInitialized());
+
+  auto *block = inst->getParent();
+  auto blockLiveness = getBlockLiveness(block);
+  if (blockLiveness == PrunedLiveBlocks::Dead)
+    return false;
+
+  bool isLive = blockLiveness == PrunedLiveBlocks::LiveOut;
+
+  if (isLive && !asImpl().isDefBlock(block))
+    return true;
+
+  // Check if instruction is between a last use and a definition
+  for (SILInstruction &it : llvm::reverse(*block)) {
+    // the def itself is not within the boundary, so cancel liveness before
+    // matching 'inst'.
+    if (asImpl().isDef(inst)) {
+      isLive = false;
+    }
+    if (&it == inst) {
+      return isLive;
+    }
+    if (!isLive && isInterestingUser(&it)) {
+      isLive = true;
+    }
+  }
+  llvm_unreachable("instruction must be in its parent block");
+}
+
+template <typename LivenessWithDefs>
+bool PrunedLiveRange<LivenessWithDefs>::areUsesWithinBoundary(
+    ArrayRef<Operand *> uses, DeadEndBlocks *deadEndBlocks) const {
+  assert(asImpl().isInitialized());
+
+  auto checkDeadEnd = [deadEndBlocks](SILInstruction *inst) {
+    return deadEndBlocks && deadEndBlocks->isDeadEnd(inst->getParent());
+  };
+  for (auto *use : uses) {
+    auto *user = use->getUser();
+    if (!asImpl().isWithinBoundary(user) && !checkDeadEnd(user))
+      return false;
+  }
+  return true;
+}
+
+template <typename LivenessWithDefs>
+bool PrunedLiveRange<LivenessWithDefs>::areUsesOutsideBoundary(
+    ArrayRef<Operand *> uses, DeadEndBlocks *deadEndBlocks) const {
+  assert(asImpl().isInitialized());
+
+  auto checkDeadEnd = [deadEndBlocks](SILInstruction *inst) {
+    return deadEndBlocks && deadEndBlocks->isDeadEnd(inst->getParent());
+  };
+  for (auto *use : uses) {
+    auto *user = use->getUser();
+    if (asImpl().isWithinBoundary(user) || checkDeadEnd(user))
+      return false;
+  }
+  return true;
+}
+
+template <typename LivenessWithDefs>
+void PrunedLiveRange<LivenessWithDefs>::findBoundariesInBlock(
+    SILBasicBlock *block, bool isLiveOut,
+    PrunedLivenessBoundary &boundary) const {
+  assert(asImpl().isInitialized());
+
+  bool isLive = isLiveOut;
+  bool isDefBlock = asImpl().isDefBlock(block);
+  SILInstruction *nextDef = nullptr;
+  SILInstruction *searchPos = block->getTerminator();
+  while (searchPos) {
+    if (isLive) {
+      nextDef = asImpl().findPreviousDef(searchPos, nextDef);
+      if (!nextDef)
```

Comment:
When there's only a single def (SSAPrunedLiveness), does continuing the search above that single def result in visiting instructions which can't be on the boundary?

Reply:
No, the main goal of this design was to avoid visiting any extra instructions for the common SSAPrunedLiveness case. That's why `findPreviousDef` is delegated to the subclass. Does `SSAPrunedLiveness::findPreviousDef` look right?

Reply:
In an example like
```
entry:
    ...
    instructions
    ...
    %def = apply
    ...
    apply(%def) // last use
    ...
    %retval = tuple ()
    return %retval

```

It looks like we'll do the following:
1. set `isLive = false`, `isDefBlock = true`, `nextDef = nullptr`
2. advance searchPos from the end of the block toward the front.  for the first instructions, `isLive` will be false, and since `isDefBlock` is true, we'll check whether instructions are the def.  none of them will be.
3. when searchPos becomes `apply(%def)`; `isInterestingUser` will return something other than NonUser so we'll set `isLive = true`
4. enter the `isLive` condition.  call `findPreviousDef(theLastUse, nullptr)`.  we'll get back `%def = apply`.
5. set `searchPos` to `nextDef`.  set `isLive` to false.
6. keep iterating over the instructions above `%def = apply`.

At least it looks like that's what I'm seeing in the debugger for the following:

```
class K {}

sil [ossa] @getK : $@convention(thin) () -> @owned K

sil [ossa] @borrowK : $@convention(thin) (@guaranteed K) -> ()

sil [ossa] @takeK : $@convention(thin) (@owned K) -> ()

sil [ossa] @f : $() -> () {
  %getK = function_ref @getK : $@convention(thin) () -> @owned K
  %borrowK = function_ref @borrowK : $@convention(thin) (@guaranteed K) -> ()
  %takeK = function_ref @takeK : $@convention(thin) (@owned K) -> ()
  %k1 = apply %getK() : $@convention(thin) () -> @owned K
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  %k2 = apply %getK() : $@convention(thin) () -> @owned K
  debug_value [trace] %k2 : $K
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %takeK(%k2) : $@convention(thin) (@owned K) -> ()
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %takeK(%k1) : $@convention(thin) (@owned K) -> ()
  %retval = tuple ()
  return %retval : $()
}
```

Reply:
@nate-chandler right! I'm glad you spotted that because I changed my mind after writing it.

/cc @eeckstein 
I intentionally continued looking for a liveness boundary in the def block handle to handle self loops:

```
bb1(%phi : @owned $K):
  apply %borrowK(%phi)
  destroy_value %phi : $K
  %k = apply %getK() : ()-> @owned $K
  br bb1(%k : $K)
```

I was concerned that, when `bb1` becomes unreachable, the `%phi` will be immediately removed by a simple peephole:

```
bb1:
  borrowK(%k) : $K
  %k = apply %getK() : ()-> @owned $K
  br bb1
```

But, I was never able to come up with a way to test this. Our SIL verifier claims that it does not meet SSA dominance requirements. In theory, it does meet SSA requirements because unreachable blocks are in fact dominated by everything in the function. In fact, introducing a meaningless branch makes it valid SIL again:

```
bb1:
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  br bb2

bb2:
  %k1 = apply %getK() : $@convention(thin) () -> @owned K
  br bb1
```

But we care about the case where the def and use are in the same block. Since we can't test this, I came up (or rediscovered) a new rule:

_Redundant phis can only be eliminated after first eliminating all unreachable code_

I'll add a comment about that optimizer rule and make this liveness check more efficient.

Reply:
@nate-chandler @meg-gupta 
Now that I realize we don't need to guard against self-loops with SSA liveness, a big pile of complexity that I added to deal with that corner case goes away:
https://github.com/apple/swift/pull/61371/commits/281b0ce32ab6e06831bcb053f0fc62c78ef670e2

Reply:
Hooray--thank you!